### PR TITLE
WIP: Number of cards

### PR DESCRIPTION
**Pull request by BlackDark**
## Description

See #1474
Updates filtering algorithms.

## Related Issue

#1474 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change) (Maybe? Cause of the changes in filtering)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: client/js/views/list_view.js [hunk 1]
```diff
@@ -70,6 +70,11 @@ App.ListView = Backbone.View.extend({
             this.model.collection.board.cards.bind('change:is_archived', this.renderCardsCollection);
             this.model.collection.board.cards.bind('change:comment_count', this.renderCardsCollection);
             this.model.collection.board.cards.bind('change:list_id', this.renderCardsCollection);
+            this.model.collection.board.cards.bind('change:is_filtered', function(e){
+                // TODO not synchronized models here?
+                console.log(e.get('is_filtered'));
```
Comment (BlackDark) [inline client/js/views/list_view.js]: @rrjanbiah Do you know why in the list_view context the cards are not really in sync with other components?
For example the same construct is in card_view and there the attribute is correctly applied.

Comment (BlackDark) [reply client/js/views/list_view.js]: Always instantiating jQuery Selectors is slow (see http://vanilla-js.com/ or https://jsperf.com/).
The big problem of jQuery usage is generating spagetti code (for example filtering in card_view and header).

So i am not sure if there are performance benefits in using jquery for manipulation of the DOM.

Comment (BlackDark) [reply client/js/views/list_view.js]: Thanks for your feedback.

Why do we have to render all cards even if they would be filtered? In this case i would implement the render function to skip the rendering of the card. As a bonus (compared to the current jQuery implementation) we would reduce the size of the DOM and increase performance where we do not need all cards shown.

I really think that it doesn't affect the performance at all and we could test that before using it in production.
My question over a month ago was the problem of outsynched models. This means that changing values in the card model does not really affect the card models in the list model (and vice-versa). 
Do you decoupled the model structures somewhere from each other? From my understanding an used model should be synched through the whole web app.

Comment (sridhar391) [reply client/js/views/list_view.js]: In `board_header_view.js` line 1841 `cardFilter function()`, you have set `is_filtered` attributes for the `cards  in the board model` 

In `list_view.js` line 1187 `renderCardNumbers function()`,  you have `filtered the cards from list model`.

 If you want to filter the cards from the board model 
In  `list_view.js` `renderCardNumbers function()`
- In line 1186 add `var self = this;`
- In line 1187  change it from 
`var filteredElements = this.model.cards.filter(function(card) {`  to  `var filteredElements = this.model.collection.board.cards.filter(function(card) {`  and please provide feedback after tried this change of code
- In line 1188 change it from
` return card.get('is_archived') !== 1 && card.get('is_filtered') === false;`  to  `return card.get('is_archived') !== 1 && card.get('list_id') === parseInt(self.model.attributes.id) && card.get('is_filtered') === false;`

please provide feedback after tried this change of codes

Comment (BlackDark) [reply client/js/views/list_view.js]: Thx seems to work for triggering the rerendering of card numbers after filter application.

Currently i do not have the time (and priority) to further develop on the counter for the rows. Maybe if you want to can continue on the code base or stay/learn from ideas for the future restructuring of the client.

Comment (sridhar391) [reply client/js/views/list_view.js]: Thanks for your contribution, we will continue the code change and will  merge the changes

**Comment by sridhar391** [comment]
@BlackDark & @rrjanbiah ,

History of conversation happened in code review was hidden. So for sake of the record, we have posted here as a comment.

-----------

From @BlackDark,

@rrjanbiah Do you know why in the list_view context the cards are not really in sync with other components?
For example, the same construct is in card_view and there the attribute is correctly applied.

-----------

From @sridhar391,

@BlackDark Actually we're using the show and hide class to filter cards with toggle operation for effective time management. Anyway, we'll consider your request. We'll check with our team and let you update

-----------

From @BlackDark,

Always instantiating jQuery Selectors is slow (see http://vanilla-js.com/ or https://jsperf.com/).
The big problem of jQuery usage is generating spagetti code (for example filtering in card_view and header).

So i am not sure if there are performance benefits in using jquery for manipulation of the DOM.

-----------

From @sridhar391,

While having more than 100 cards in the board, if we filter using card backbone model, all the cards in the board page will be rendered and the backbone will consume lot of memory for this process than the jQuery. so, that only we use the jQuery Selectors.

If you strongly think it will not affect any performance, then if you need any help to change the code to use card backbone model or shall we do those changes?

-----------

From @BlackDark,

Thanks for your feedback.

Why do we have to render all cards even if they would be filtered? In this case i would implement the render function to skip the rendering of the card. As a bonus (compared to the current jQuery implementation) we would reduce the size of the DOM and increase performance where we do not need all cards shown.

I really think that it doesn't affect the performance at all and we could test that before using it in production.
My question over a month ago was the problem of outsynched models. This means that changing values in the card model does not really affect the card models in the list model (and vice-versa).
Do you decoupled the model structures somewhere from each other? From my understanding an used model should be synched through the whole web app.

-----------

From @sridhar391,

We will check and update you soon

-----------

From @sridhar391,

In `board_header_view.js` line 1841 `cardFilter function(),` you have set `is_filtered` attributes for the cards in the board model

In `list_view.js` line 1187 `renderCardNumbers function(),` you have filtered the cards from list model.

If you want to filter the cards from the board model in `list_view.js` file `renderCardNumbers function()`

- In line 1186 add `var self = this;`
- In line 1187 change it from
  `var filteredElements = this.model.cards.filter(function(card) {` to `var filteredElements = this.model.collection.board.cards.filter(function(card) {`
- In line 1188 change it from `return card.get('is_archived') !== 1 && card.get('is_filtered') === false;` to `return card.get('is_archived') !== 1 && card.get('list_id') === parseInt(self.model.attributes.id) && card.get('is_filtered') === false;`

Please provide feedback after tried this change of codes.

-----------

From @BlackDark,

Thx seems to work for triggering the rerendering of card numbers after filter application.

Currently i do not have the time (and priority) to further develop on the counter for the rows. Maybe if you want to can continue on the code base or stay/learn from ideas for the future restructuring of the client.

-----------

From @sridhar391,

Thanks for your contribution, we will continue the code change and will merge the changes

**Comment by sridhar391** [comment]
We have added the code changes for showing the number of cards in this commit https://github.com/RestyaPlatform/board/commit/e698399c005b5179360fb95dfcb076bdf0b3b88b and the code is merged in this pull request https://github.com/RestyaPlatform/board/pull/2310

**Pull request closed without merging**